sdk/storage/azure-storage-blob/azure/storage/blob/_download.py

```diff
@@ -205,17 +204,17 @@ def _download_chunk(self, chunk_start, chunk_end):
                         download_stream_current=self.progress_total,
                         **self.request_options
                     )
-                except HttpResponseError as error:
-                    process_storage_error(error)
-
-                try:
-                    chunk_data = process_content(response, offset[0], offset[1], self.encryption_options)
                     retry_active = False
-                except (requests.exceptions.ChunkedEncodingError, requests.exceptions.ConnectionError) as error:
```

Comment:
I think this is trying to catch exceptions raised by `self.client.download` (line 200). Same idea below.

Reply:
I had a look at this and I'm not sure moving the retry logic away from `process_content` was correct here. The call to `client.download` will actually returning an `IO` object that can stream the response from the service (if the download I big enough). This means inside `process_content` we will actually be making additional requests that I believe this retry logic is trying to catch. The requests will happen here:
https://github.com/Azure/azure-sdk-for-python/blob/b0f9e02b2d597a57dcef33bb265c4b5c47ad0a85/sdk/storage/azure-storage-blob/azure/storage/blob/_download.py#L46

I don't know the full history here, but I believe at one point the additional requests made during response streaming did not pass through our normal pipeline and therefore did not get our normal exception handling or retry policies. This would make sense why additional retry logic was added here and why raw `requests` and `aiohttp` exceptions were being caught. I'm honestly not sure if that is still the case though or if there was a change to make response streaming use the normal pipeline.

If response streaming now uses the normal pipeline, we may not even need the retries here at all and may actually be tripling our retries if we start catching the core exceptions.

@lmazuel, @xiangyan99, @iscai-msft do any of you know if response streaming uses the standard pipeline now? The generated code accepts the pipeline as a param now so that would seem to indicate it does?
https://github.com/Azure/azure-sdk-for-python/blob/b0f9e02b2d597a57dcef33bb265c4b5c47ad0a85/sdk/storage/azure-storage-blob/azure/storage/blob/_generated/operations/_blob_operations.py#L1757

Reply:
I think getting rid of the retries might be the way to go. When I debug through the following code, each iteration (except for the first one) in the for loop goes into the pipeline.

```python
    account_client = BlobServiceClient(ENDPOINT, KEY)
    blob_client = BlobClient(
        account_url=account_client.url,
        container_name="tsjinxuanstorage2",
        blob_name="random",
        credential=account_client.credential,
        max_single_get_size=1,
        max_chunk_get_size=1,
    )

    downloader = blob_client.download_blob()
    for chunk in downloader.chunks():
        print(chunk)
```

Is this what you mean by an `IO` object that streams data? If so, I think you are right. The pipeline itself has a [retry policy]( https://github.com/Azure/azure-sdk-for-python/blob/1486a398c39cf9e2f1b7aa938a3e7e52f55fd6e1/sdk/storage/azure-storage-blob/azure/storage/blob/_shared/base_client.py#L243) so my change would indeed be tripling the number of retries. Also, this means that we should update #24965 accordingly.

Reply:
@Stevenjin8 this code will actually make a series of regular GET (`self.cleint.download`) requests that each ask for 1 Byte so response streaming would not actually kick in here. All of the data would come in the initial download call and no additional requests would happen in `process_content` AFAIK. The pipeline being triggered while debugging will come from each explicit call to download.

I'm not entirely sure what triggers response streaming, but you can see we are using it here. I guess I'd assume it would be triggered by a certain size payload from the service?
https://github.com/Azure/azure-sdk-for-python/blob/0a20bfba26ac5f43dcecb203311df63e3abe24c6/sdk/storage/azure-storage-blob/azure/storage/blob/_generated/operations/_blob_operations.py#L1702

It's sort of tough to tell if the code in `process_content` is actually sending requests but possibly if you try downloading a large payload in a single request (up to 32 MiB we expect only one request by default) and see if the pipeline gets hit multiple times., that may help to shed some light.

Reply:
@jalauzon-msft I think I see what you mean, but I also can't seem to get streaming to run. I'll talk to @lmazuel about it tomorrow in our 1:1.

Reply:
@jalauzon-msft I'm a bit lost and I think it might be better to leave things as they are until we figure out exactly what is going on with streaming (I can make an issue).

https://github.com/Azure/azure-sdk-for-python/blob/b0f9e02b2d597a57dcef33bb265c4b5c47ad0a85/sdk/storage/azure-storage-blob/azure/storage/blob/_download.py#L46

What I had in mind is to keep most of the code as it is and just replace the `requests` exceptions with `azure.core` exceptions. But if that line is not using the pipeline, then I would be tripling the retry mechanism. If that line is not using the pipeline, then I would be getting rid of retries all together.

Reply:
Given your investigation, I think this is the best course of action. It seems the response streaming no longer throws these `requests` errors and therefore, currently, our retry logic here is not doing anything. Additionally, looking into the implementation of `StreamDownloadGenerator` it seems to be reading the raw response, meaning it's likely not using our normal retry pipeline so I don't think we'll be tripling the retry count.